### PR TITLE
IR: factor identical slices in multi dst effects

### PR DESCRIPTION
**Pull request by serpilliere**
`get_modified_slice` can send back multiple identical slices.
Those slices need to be factorised for next steps.

### CODE REVIEW

File: miasm2/ir/ir.py [hunk 1]
```diff
@@ -176,6 +176,9 @@ def merge_multi_affect(self, affect_list):
             e_colision = reduce(
                 lambda x, y: x + y, [e.get_modified_slice() for e in expr_list])
 
+            # Factorize identic collitions
+            e_colision = list(set(e_colision))
+
             # Sort interval collision
             known_intervals = sorted([(x[1], x[2]) for x in set(e_colision)])
```
Comment (p-l-) [inline miasm2/ir/ir.py]: `set()` is useless now here.

File: miasm2/ir/ir.py [hunk 2]
```diff
@@ -176,6 +176,9 @@ def merge_multi_affect(self, affect_list):
             e_colision = reduce(
                 lambda x, y: x + y, [e.get_modified_slice() for e in expr_list])
 
+            # Factorize identic collitions
+            e_colision = list(set(e_colision))
```
Comment (p-l-) [inline miasm2/ir/ir.py]: Can't you keep a `set()` instead of creating another list?

File: miasm2/ir/ir.py [hunk 3]
```diff
@@ -176,6 +176,9 @@ def merge_multi_affect(self, affect_list):
             e_colision = reduce(
                 lambda x, y: x + y, [e.get_modified_slice() for e in expr_list])
```
Comment (p-l-) [inline miasm2/ir/ir.py]: This creates two lists (one as the output of `reduce()`, one by list comprehension.

What do you think of replacing lines 176 to 180 by something like:

```python
e_colision = reduce(lambda x, y: set(x).union(y), (e.get_modified_slice() for e in expr_list))
```

If you really need a list use `e_colision = list(reduce([...]))`. And anyway drop the `set()` in `known_intervals = [...]`

**Comment by serpilliere** [comment]
I have updated the code.
By the way, there was a mistake in the code:
```python
missing_i = get_missing_interval(known_intervals, 0, e.src.size)
```

`e` was referencing a variable generated in a comprehension list.

File: miasm2/ir/ir.py [hunk 4]
```diff
@@ -173,21 +173,20 @@ def merge_multi_affect(self, affect_list):
                 continue
 
             # Find collision
-            e_colision = reduce(
-                lambda x, y: x + y, [e.get_modified_slice() for e in expr_list])
-
+            e_colision = reduce(lambda x, y: set(x).union(y),
+                                (e.get_modified_slice() for e in expr_list))
```
Comment (p-l-) [inline miasm2/ir/ir.py]: By bad, it's better written as this:

```Python
e_colision = reduce(lambda x, y: x.union(y), (e.get_modified_slice() for e in expr_list), set())
```

This is a tiny fix but I think it's better as it avoids to call the `set()` constructor on a `set` object for each element from the generator. Sorry I didn't write it like that in the first place.

The rest of the patch looks good to me.

**Comment by serpilliere** [comment]
Modification done.

File: miasm2/ir/ir.py [hunk 5]
```diff
@@ -173,21 +173,21 @@ def merge_multi_affect(self, affect_list):
...
-            known_intervals = sorted([(x[1], x[2]) for x in set(e_colision)])
+            known_intervals = sorted([(x[1], x[2]) for x in e_colision])
 
             # Fill with missing data
-            missing_i = get_missing_interval(known_intervals, 0, e.src.size)
+            missing_i = get_missing_interval(known_intervals, 0, dst.size)
 
-            rest = [(m2_expr.ExprSlice(dst, r[0], r[1]), r[0], r[1])
-                    for r in missing_i]
+            remaining = [(m2_expr.ExprSlice(dst, *interval),
```
Comment (p-l-) [inline miasm2/ir/ir.py]: Unless I'm wrong you do not use `remaining` after line 199 where a generator would be enough and use less memory. Just replace `[...]` by `(...)`:

```Python
remaining = ((m2_expr.ExprSlice(dst, *interval), interval[0], interval[1]) for interval in missing_i)

**Comment by serpilliere** [comment]
done.

File: miasm2/ir/ir.py [hunk 6]
```diff
@@ -173,21 +173,22 @@ def merge_multi_affect(self, affect_list):
...
-            known_intervals = sorted([(x[1], x[2]) for x in set(e_colision)])
+            known_intervals = sorted([(x[1], x[2]) for x in e_colision])
 
             # Fill with missing data
-            missing_i = get_missing_interval(known_intervals, 0, e.src.size)
+            missing_i = get_missing_interval(known_intervals, 0, dst.size)
 
-            rest = [(m2_expr.ExprSlice(dst, r[0], r[1]), r[0], r[1])
-                    for r in missing_i]
+            remaining = remaining = ((m2_expr.ExprSlice(dst, *interval),
```
Comment (p-l-) [inline miasm2/ir/ir.py]: I know you're going to hate me but... you have one `remaining = ` more than needed here. Sorry mate, I owe you a good beer for that PR.

**Comment by serpilliere** [comment]
done.

**Pull request merged**